File: [K]graph/[K]graph-union-find/721-accounts-merge.py
class UnionFind:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.rank = [0 for _ in range(n)]

    def find(self, p):
        while p != self.parent[p]:
            self.parent[p] = self.parent[self.parent[p]]
            p = self.parent[p]
        return p
        
    def union(self, p, q):
        root_p = self.find(p)
        root_q = self.find(q)
        if root_p == root_q:
            return
        if self.rank[root_p] > self.rank[root_q]:
            self.parent[root_q] = root_p
        elif self.rank[root_p] < self.rank[root_q]:
            self.parent[root_p] = root_q
        else:
            self.parent[root_p] = root_q
            self.rank[root_q] += 1
        
from collections import defaultdict
class Solution:
    def accountsMerge(self, accounts: List[List[str]]) -> List[List[str]]:
        n = len(accounts)
        uf = UnionFind(n)
        email_id = {}
        for i, info in enumerate(accounts):
            for j in range(1, len(info)):
                if info[j] in email_id:
                    uf.union(i, email_id[info[j]])
                else:
                    email_id[info[j]] = i

        id_emails = defaultdict(set)
        for email, id in email_id.items():
            id_emails[uf.find(id)].add(email)
        
        res = []
        for id, emails in id_emails.items():
            info = [accounts[id][0]]
            info.extend(sorted(emails))
            res.append(info)
        return res
    
# time O(mlogm + n + m), due to sort
# space O(m+n), due to hashmap, m is the number of mails, n is the number of accounts
# using graph and union find and sort

File: [K]graph/[K]graph-union-find/721-accounts-merge.java
class UnionFind {
    int[] parents;
    int[] rank;

    public UnionFind(Integer n) {
        parents = new int[n];
        rank = new int[n];
        for (int i = 0; i < n; i++) {
            parents[i] = i;
            rank[i] = 0;
        }
    }

    public Integer find(Integer p) {
        while (p != parents[p]) {
            parents[p] = parents[parents[p]];
            p = parents[p];
        }
        return p;
    }

    public void union(Integer p, Integer q) {
        Integer rootP = find(p);
        Integer rootQ = find(q);
        if (rootP == rootQ) {
            return;
        }
        if (rank[rootP] > rank[rootQ]) {
            parents[rootQ] = rootP;
        } else if (rank[rootP] < rank[rootQ]) {
            parents[rootP] = rootQ;
        } else {
            parents[rootP] = rootQ;
            rank[rootQ] += 1;
        }
    }
}

class Solution {
    public List<List<String>> accountsMerge(List<List<String>> accounts) {
        UnionFind uf = new UnionFind(accounts.size());
        HashMap<String, Integer> emailToId = new HashMap<>();
        for (int i = 0; i < accounts.size(); i++) {
            for (int j = 1; j < accounts.get(i).size(); j++) {
                String mail = accounts.get(i).get(j);
                if (emailToId.containsKey(mail)) {
                    uf.union(emailToId.get(mail), i);
                } else {
                    emailToId.put(mail, i);
                }
            }
        }

        HashMap<Integer, Set<String>> idToEmails = new HashMap<>();
        for (int i = 0; i < accounts.size(); i++) {
            for (int j = 1; j < accounts.get(i).size(); j++) {
                String mail = accounts.get(i).get(j);
                Integer id = uf.find(i);
                idToEmails.computeIfAbsent(id, k -> new HashSet<>()).add(mail);
            }
        }

        List<List<String>> res = new ArrayList<>();
        for (Map.Entry<Integer, Set<String>> entry: idToEmails.entrySet()) {
            Integer id = entry.getKey();
            Set<String> mails = entry.getValue();
            ArrayList<String> info = new ArrayList<>();
            info.addAll(mails);
            Collections.sort(info);
            info.add(0, accounts.get(id).get(0));
            res.add(info);
        }
        return res;
    }
}

// time O(mlogm + n + m), due to sort
// space O(m+n), due to hashmap, m is the number of mails, n is the number of accounts
// using graph and union find and sort